### DISCUSSION TITLE
"parent_device" property seems to be unseen by server at REST HTTP POST

### DISCUSSION THREAD

**Discussion by ViktorTelecom** [netbox-community/netbox#11713] [Q&A] [answered]
Dear community,

I have problems with uploading DCIM configuration when I need to populate my already uploaded parent chassis device with child devices. My HTTP POST message is the following. I collected it with TCP Dump.

```
HTTP: POST /api/dcim/devices/ HTTP/1.1
Authorization: Token gd7jjgsdaqdhfghdl8ydshdjhsdhs
Transfer-Encoding: chunked
traceparent: 00-cb642d0a00e80a058297c09f105ede82-1078be4dbf8e0b6b-00
Content-Type: application/json; charset=utf-8


{
  "name": "TestCard (1)",
  "device_type": {
    "model": "TestCard"
  },
  "device_role": {
    "name": "Linear Card"
  },
  "site": {
    "slug": "main-site"
  },
  "rack": {
    "name": "room 206, place 04",
    "site": {
      "slug": "main-site"
    }
  },
  "position": null,
  "face": null,
  "status": "planned",
  "comments": "",
  "parent_device": {
    "id": 254,
    "name": "QPL-WEM-1",
    "device_bay": {
      "name": "Bay-1",
      "device": {
        "id": 254
      }
    }
  }
}
```

The following is the server reply. Also captured with tcpdump.

```
`HTTP/1.1 201 Created
Server: nginx/1.18.0 (Ubuntu)
Date: Thu, 09 Feb 2023 11:39:15 GMT
Content-Type: application/json
Content-Length: 1636
Connection: keep-alive
Location: http://netbox.example.com/api/dcim/devices/258/
Vary: Accept, Cookie, Origin
Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
API-Version: 3.2
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000; includeSubDomains

{
  "id": 258,
  "url": "http://netbox.example.com/api/dcim/devices/258/",
  "display": "TestCard  (1)",
  "name": "TestCard  (1)",
  "device_type": {
    "id": 19,
    "url": "http://netbox.example.com/api/dcim/device-types/19/",
    "display": "TestCard",
    "manufacturer": {
      "id": 6,
      "url": "http://netbox.example.com/api/dcim/manufacturers/6/",
      "display": "TestVendor",
      "name": "TestVendor",
      "slug": "testvendor"
    },
    "model": "TestCard",
    "slug": "testcard"
  },
  "device_role": {
    "id": 16,
    "url": "http://netbox.example.com/api/dcim/device-roles/16/",
    "display": "Linear Card",
    "name": "Linear Card",
    "slug": "linear-card"
  },
  "tenant": null,
  "platform": null,
  "serial": "",
  "asset_tag": null,
  "site": {
    "id": 81,
    "url": "http://netbox.example.com/api/dcim/sites/81/",
    "display": "Main Site",
    "name": "Main Site",
    "slug": "main-site"
  },
  "location": null,
  "rack": {
    "id": 154,
    "url": "http://netbox.example.com/api/dcim/racks/154/",
    "display": "room 206, place 04",
    "name": "room 206, place 04"
  },
  "position": null,
  "face": null,
  "parent_device": null,
  "status": {
    "value": "planned",
    "label": "Planned"
  },
  "airflow": null,
  "primary_ip": null,
  "primary_ip4": null,
  "primary_ip6": null,
  "cluster": null,
  "virtual_chassis": null,
  "vc_position": null,
  "vc_priority": null,
  "comments": "",
  "local_context_data": null,
  "tags": [],
  "custom_fields": {},
  "config_context": {},
  "created": "2023-02-09T11:39:15.551540Z",
  "last_updated": "2023-02-09T11:39:15.551551Z"
}
```

We see that almost all introduced properties had been successfully accepted and processed. Only property "parent_device" was somehow ignored. I tried to send data of any type under "parent_device" but there was no reaction from the server. Even no error message. As if this property does not exist at all on server.

In web console I can see the created module on the correct site and in the correct rack so it it can be manually installed to the parent device "device-bay". But this does not go right with API.

What do I do wrong? It seems to be an obvious scenario.

When I use modules instead I successfully install modules in module-bays of parent device. But modules do not allow the same port name on different modules on one device. So I use child devices.

**Comment by jsenecal** [comment, marked as answer]
Thats because `parent_device` is a nested serializer: https://github.com/netbox-community/netbox/blob/19da92b5108772288d960d350d7c7083851359a0/netbox/dcim/api/serializers.py#L663-L672
I don't think it was ever intended to get written to however.

You need to find the `device_bay` id and then add your newly created device there: `put /dcim/device-bays/{id}/` if I'm not mistaken...

**Reply by jsenecal** [reply]
I think we need to add `read_only=True` there: https://github.com/netbox-community/netbox/blob/19da92b5108772288d960d350d7c7083851359a0/netbox/dcim/api/serializers.py#L649

**Reply by ViktorTelecom** [reply]
@jsenecal It works! Thank you for assistance.

```
PUT /api/dcim/device-bays/ HTTP/1.1
............
[
{
  "id": 132,
  "name": "Bay-1",
  "device": 339,
  "installed_device": {
    "id": 340
  }
}
]


HTTP/1.1 200 OK
............

[
  {
    "id": 132,
    "url": "http://netbox.example.com/api/dcim/device-bays/132/",
    "display": "Bay-1",
    "device": {
      "id": 339,
      "url": "http://netbox.example.com/api/dcim/devices/339/",
      "display": "QPL-WEM-1",
      "name": "QPL-WEM-1"
    },
    "name": "Bay-1",
    "label": "",
    "description": "",
    "installed_device": {
      "id": 340,
      "url": "http://netbox.example.com/api/dcim/devices/340/",
      "display": "TestCard (1)",
      "name": "TestCard  (1)"
    },
    "tags": [],
    "custom_fields": {},
    "created": "2023-02-10T09:36:32.792055Z",
    "last_updated": "2023-02-10T09:36:33.228140Z"
  }
]
```

**Comment by mtinberg** [comment]
The format used to POST/PUT/PATCH data isn't the same as data is returned, eg foreign key fields are often returned with details like name, id, display, url, etc. but are written as just the id number, eg installedDevice: 1234.  or sometimes they are updated on a different data model entirely, eg. IP Address is associated with Interface on the IP Address model not the Interface model, but is available on Interface and Device. Take a look at the Swagger/OpenAPI docs to see what the field names and expected values are for the field you are trying to update.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by ViktorTelecom** [reply]
Hi, Mark!

I use swagger data but this information was not useful for my task. Because "installed_device" property can be found in several places and swagger does not show which exact object is the true owner of the property while others just reflect it. So I need to try everything or ask the community.
